README
SWIPE
Smith-Waterman database searches with Inter-sequence Parallel Execution

Copyright (C) 2008-2021 Torbjorn Rognes, University of Oslo,
Oslo University Hospital and Sencel Bioinformatics AS

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.

Contact: Torbjorn Rognes <torognes@ifi.uio.no>, 
Department of Informatics, University of Oslo, 
PO Box 1080 Blindern, NO-0316 Oslo, Norway

--

SWIPE version 2 incorporates a number of new features compared to the initial
release, most notably hit statistics, alignments, as well as support for
nucleotide sequences, large databases and MPI.

The software is distributed from http://github.com/torognes/swipe
Older versions are also available from http://dna.uio.no/swipe/

Binary executables are available for Linux and Mac (both 64-bit).

SWIPE is distributed under the GNU Affero General Public License, version 3.
See the file COPYING for details. SWIPE includes some public domain code for
computing alignment statistics extracted from the BLAST software by the
National Center for Biotechnology Information (NCBI).

Please cite: Rognes T (2011) Faster Smith-Waterman database searches
with inter-sequence SIMD parallelisation. BMC Bioinformatics 12, 221.

Usage: ./[mpi]swipe [OPTIONS]
  -h, --help                 show help
  -d, --db=FILE              sequence database base name (required)
  -i, --query=FILE           query sequence filename (stdin)
  -M, --matrix=NAME/FILE     score matrix name or filename (BLOSUM62)
  -q, --penalty=NUM          penalty for nucleotide mismatch (-3)
  -r, --reward=NUM           reward for nucleotide match (1)
  -G, --gapopen=NUM          gap open penalty (11)
  -E, --gapextend=NUM        gap extension penalty (1)
  -v, --num_descriptions=NUM sequence descriptions to show (250)
  -b, --num_alignments=NUM   sequence alignments to show (100)
  -e, --evalue=REAL          maximum expect value of sequences to show (10.0)
  -k, --minevalue=REAL       minimum expect value of sequences to show (0.0)
  -c, --min_score=NUM        minimum score of sequences to show (1)
  -u, --max_score=NUM        maximum score of sequences to show (inf.)
  -a, --num_threads=NUM      number of threads to use [1-256] (1)
  -m, --outfmt=NUM           output format [0,7-9=plain,xml,tsv,tsv+] (0)
  -I, --show_gis             show gi numbers in results (no)
  -p, --symtype=NAME/NUM     symbol type/translation [0-4] (1)
  -S, --strand=NAME/NUM      query strands to search [1-3] (3)
  -Q, --query_gencode=NUM    query genetic code [1-23] (1)
  -D, --db_gencode=NUM       database genetic code [1-23] (1)
  -x, --taxidlist=FILE       taxid list filename (none)
  -N, --dump=NUM             dump database [0-2=no,yes,split headers] (0)
  -H, --show_taxid           show taxid etc in results (no)
  -o, --out=FILE             output file (stdout)
  -z, --dbsize=NUM           set effective database size (0)

Defaults are indicated in parentheses in the option list above. NUM means an
integer, while REAL means a floating point number (possibly in scientific
notation). NAME is a string or integer, while FILE is a filename. For databases,
FILE is the base name of the database files.

Most features and options are similar and compatible with NCBI
BLAST.

All scores (e.g. for options -c and -u) are raw alignment scores
(not bit scores) unless otherwise noted.

Please note that SWIPE does not employ composition based statistics (BLAST
option "-C"), but uses the standard Karin-Altschul statistics. Also, sequence
filtering of low complexity regions (BLAST option "-F") is not implemented.
Please pre-filter your query. SWIPE does not use the new sequence length
adjustment statistics (FSC) introduced in BLAST 2.2.26.

The BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80, BLOSUM90, PAM30, PAM70 and PAM250
matrices are "built-in". Other matrices may be specified with a file name.

The query and database sequence type is specified with the "-p" option using
either a numeric or text argument according to the table below:

Numeric String  Query  Database Comparisons
0       blastn  NT     NT       Direct + reverse complementary
1       blastp  AA     AA       Direct
2       blastx  NT     AA       Translated query (6 frames)
3       tblastn AA     NT       Translated database (6 frames)
4       tblastx NT     NT       Translated query and database (6x6 frames)

A plain text file with a list of taxid numbers, one per line, may be
specified with the "-x" option to search only sequences from the
specified organisms.

SWIPE accepts only databases prepared using the formatdb or makeblastdb
tools that are distributed together with NCBI BLAST and BLAST+. They
can be downloaded at ftp://ftp.ncbi.nlm.nih.gov/blast/executables/
from NCBI. For makeblastdb, please use the "-blastdb_version 4" option
to make databases that are compatible with SWIPE.

The SWIPE distribution includes executable binaries for 64-bit Linux and
Mac. SWIPE may be compiled from sources using either the GNU g++ compiler,
the LLVM Clang compiler or the Intel icpc compiler. SWIPE is marginally
faster when compiled with the Intel compiler than with the GNU compiler.

SWIPE will only run on processors with the SSE2 instruction set and
runs best on processors with the SSSE3 instruction set. Please see
http://en.wikipedia.org/wiki/SSSE3 for a list of processors with this
instruction set.

The enclosed file scoring.pdf contains a table with the scoring system
parameters that will enable E-value statistics to be computed.

Examples of how to run SWIPE:

To run SWIPE with one or more protein sequence queries in the
FASTA-formatted file named "query.fsa" against the protein
database with base name "swissprot" using 8 threads and an expect
value threshold of 0.001:

./swipe -p 1 -i query.fsa -d swissprot -e 0.001 -a 8

To run SWIPE with a nucleotide sequence query in the file "primer.fsa"
against the nucleotide database file with base name "ecoli" using 1
thread and the default expect value of 10:

./swipe -p 0 -i primer.fsa -d ecoli

To search only the Human entries in the "est" database using a
translated search with the protein query sequence in the file
"prot.fsa":

echo 9606 > human.tax
./swipe -p 3 -i prot.fsa -d est -x human.tax

CHANGES
2021-06-28  Torbjorn Rognes  <torognes@ifi.uio.no>
        * Version 2.1.1
        * Fix for very long header strings in NCBI database files

2018-05-04  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.1.0
        * Fix for changes to NCBI database format
        * Added query description to XML output
        * Fixed compilation errors with C++11

2015-12-18  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.12
	* Fixed a bug in memory allocation that caused segfaults.
	* Fixed a minor memory leak.

2014-06-27  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.11
	* Resolved a bug that resulted in SWIPE terminating with a fatal
	  internal error in the align function. After recomputing an alignment
	  score > 32768 with 16-bit magnitude, a subsequent 16-bit score
	  computation could be wrong.
	* Reduced memory usage when a the number of alignments or results
	  asked for using the -b and -v options was higher than the possible
	  maximum number.

2014-06-25  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.10
	* Resolved an inconsistency in SWIPE with custom non-symmetric
	  matrices. This problem could result in SWIPE terminating with a
	  fatal internal error in the align function.
  
2014-01-21  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.9
	* Changed inline assembler syntax (adding register prefix) to
	  adapt to the LLVM clang compiler on Mac
	* Improved detection of CPU features (SSSE3 etc) and added
	  alternative code for the search7 function for CPUs with 
	  and without SSSE3
	* Fixed an inconsistency in the counting of 'positives', i.e.
	  similar but non-identical residues
	* Added binaries for Mac, including an openmpi version

2014-01-17  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.8
	* Removed a bug in computation of the score of alignments 
	  with a score higher than 32767
	* Removed a bug in computation of the alignment of sequences
	  that scored higher than 65535
	* No Mac executable this time, but coming soon (need to adapt
	  inline assembler syntax to the LLVM clang compiler)

2013-05-02  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.7
	* Removed a few memory leaks discovered with Valgrind

2013-05-01  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.6
	* Added some information to the README file
	* Scoring system overview in scoring.pdf
	* Code changes to avoid warnings with g++
	* Compiled on Macs too
	* Minor typo in help text corrected
	* Changed default compiler to GNU g++
	* Increased size of buffers for deflines
	* Added option (-z) to set effective length of the database

2012-08-09  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.5
	* Added option (-H) to show taxid and membership/link bits

2012-07-04  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.4
	* Fixed display of header lines with TSV format (-m 9)

2012-06-11  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.3
	* Fixed bug with option "-c" not being allowed

2012-06-08  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.2
	* Fixed bug relating to amino acids symbols "JOU*" on non-SSSE3 cpus
	* Minor output format changes / bug fixes

2012-02-24  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.1
	* hits.cc: Fixed bug in display of alignment coordinates 
          in XML and TSV formats

2012-02-01  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 2.0.0
	* Initial release of version 2 with many new features

2011-04-11  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 1.0
	* Added scoring matrices to the distribution

2011-02-25  Torbjorn Rognes  <torognes@ifi.uio.no>
	* Version 1.0
	* Initial release of the software

query.cc
/*
    SWIPE
    Smith-Waterman database searches with Inter-sequence Parallel Execution

    Copyright (C) 2008-2013 Torbjorn Rognes, University of Oslo,
    Oslo University Hospital and Sencel Bioinformatics AS

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Contact: Torbjorn Rognes <torognes@ifi.uio.no>,
    Department of Informatics, University of Oslo,
    PO Box 1080 Blindern, NO-0316 Oslo, Norway
*/

#include "swipe.h"

//   @   A   B   C   D   E   F   G   H   I   J   K   L   M   N   O
//   P   Q   R   S   T   U   V   W   X   Y   Z   [   \   ]   ^   |

char map_sound[256] = 
  {
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15,
    16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, -1, -1, -1, -1, -1,
    -1, 27, 28, 29, 30, 31, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1
  };

char map_ncbi_aa[256] = 
  {
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 25, -1, -1,  0, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1,  1,  2,  3,  4,  5,  6,  7,  8,  9, 27, 10, 11, 12, 13, 26,
    14, 15, 16, 17, 18, 24, 19, 20, 21, 22, 23, -1, -1, -1, -1, -1,
    -1,  1,  2,  3,  4,  5,  6,  7,  8,  9, 27, 10, 11, 12, 13, 26,
    14, 15, 16, 17, 18, 24, 19, 20, 21, 22, 23, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1
  };

char map_ncbi_nt4[256] = 
  {
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1,  0, -1,  1, -1, -1, -1,  2, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1,  3,  3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1,  0, -1,  1, -1, -1, -1,  2, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1,  3,  3, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1
  };

char map_ncbi_nt16[256] = 
  {
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1,  1, 14,  2, 13, -1, -1,  4, 11, -1, -1, 12, -1,  3, 15, -1,
    -1, -1,  5,  6,  8,  8,  7,  9, -1, 10, -1, -1, -1, -1, -1, -1,
    -1,  1, 14,  2, 13, -1, -1,  4, 11, -1, -1, 12, -1,  3, 15, -1,
    -1, -1,  5,  6,  8,  8,  7,  9, -1, 10, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1
  };


char ntcompl[16] = { 0, 8, 4, 12, 2, 10, 6, 14, 1, 9, 5, 13, 3, 11, 7, 15 };

char q_translate[16*16*16];
char d_translate[16*16*16];

const char * gencode_names[23] = 
  {
    "Standard Code",
    "Vertebrate Mitochondrial Code",
    "Yeast Mitochondrial Code",
    "Mold, Protozoan, and Coelenterate Mitochondrial Code and Mycoplasma/Spiroplasma Code",
    "Invertebrate Mitochondrial Code",
    "Ciliate, Dasycladacean and Hexamita Nuclear Code",
    NULL,
    NULL,
    "Echinoderm and Flatworm Mitochondrial Code",
    "Euplotid Nuclear Code",
    "Bacterial, Archaeal and Plant Plastid Code",
    "Alternative Yeast Nuclear Code",
    "Ascidian Mitochondrial Code",
    "Alternative Flatworm Mitochondrial Code",
    "Blepharisma Nuclear Code",
    "Chlorophycean Mitochondrial Code",
    NULL,
    NULL,
    NULL,
    NULL,
    "Trematode Mitochondrial Code",
    "Scenedesmus obliquus Mitochondrial Code",
    "Thraustochytrium Mitochondrial Code"
  };

const char * code[23] =
  { 
    "FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNKKSS**VVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CCWWTTTTPPPPHHQQRRRRIIMMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNKKSSSSVVVVAAAADDEEGGGG",
    "FFLLSSSSYYQQCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    NULL,
    NULL,
    "FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNNKSSSSVVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CCCWLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CC*WLLLSPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNKKSSGGVVVVAAAADDEEGGGG",
    "FFLLSSSSYYY*CCWWLLLLPPPPHHQQRRRRIIIMTTTTNNNKSSSSVVVVAAAADDEEGGGG",
    "FFLLSSSSYY*QCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FFLLSSSSYY*LCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    NULL,
    NULL,
    NULL,
    NULL,
    "FFLLSSSSYY**CCWWLLLLPPPPHHQQRRRRIIMMTTTTNNNKSSSSVVVVAAAADDEEGGGG",
    "FFLLSS*SYY*LCC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG",
    "FF*LSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG"
  };
  
char remap[] = { 2, 1, 3, 0 };
  
//                       00000000001111111111222222222233
//                       01234567890123456789012345678901
const char * sym_ncbi_nt4   = "acgt############################";
const char * sym_ncbi_nt16  = "-acmgrsvtwyhkdbn################";
const char * sym_ncbi_nt16u = "-ACMGRSVTWYHKDBN################";
const char * sym_ncbi_aa    = "-ABCDEFGHIKLMNPQRSTVWXYZU*OJ####";
const char * sym_sound      = "-ABCDEFGHIJKLMNOPQRSTUVWXYZabcde";

struct query_s query;

FILE * query_fp;
char query_line[LINE_MAX];

void query_init(const char * queryname, long symtype, long strands)
{
  if (strcmp(queryname, "-") == 0)
    query_fp = stdin;
  else
    query_fp = fopen(queryname, "r");
  
  if (!query_fp)
    fatal("Cannot open query file.");
  
  query.description = 0;
  query.dlen = 0;
  query.symtype = symtype;
  query.strands = strands;

  if (query.symtype == 5)
  {
    query.map = map_sound;
    query.sym = sym_sound;
  }
  else if ((query.symtype == 1) || (query.symtype == 3))
  {
    query.map = map_ncbi_aa;
    query.sym = sym_ncbi_aa;
  }
  else
  {
    query.map = map_ncbi_nt16;
    query.sym = sym_ncbi_nt16;
  }

  for(long s=0; s<2; s++)
  {
    query.nt[s].seq = 0;
    query.nt[s].len = 0;
    
    for(long f=0; f<3; f++)
    {
      query.aa[3*s+f].seq = 0;
      query.aa[3*s+f].len = 0;
    }
  }

  query_line[0] = 0;
  query.aa[0].seq_unmasked = 0;
  fgets(query_line, LINE_MAX, query_fp);
}

void query_free()
{
  if (query.description)
    free(query.description);
  query.description = 0;
  query.dlen = 0;

  for(long s=0; s<2; s++)
  {
    if (query.nt[s].seq)
      free(query.nt[s].seq);
    query.nt[s].seq = 0;
    query.nt[s].len = 0;
    
    for(long f=0; f<3; f++)
    {
      if (query.aa[3*s+f].seq)
	free(query.aa[3*s+f].seq);
      query.aa[3*s+f].seq = 0;
      query.aa[3*s+f].len = 0;
    }
  }
  if (mask && query.aa[0].seq_unmasked) {
    free(query.aa[0].seq_unmasked);
    query.aa[0].seq_unmasked = 0;
  }
}

void query_exit()
{
  if (query_fp != stdin)
    fclose(query_fp);

  query_free();
}

int query_read()
{
  if (!query_line[0])
    return 0;

  query_free();

  // read description

  int len = strlen(query_line);
  
  if (query_line[len-1] == '\n')
  {
    query_line[len-1] = 0;
    len--;
  }

  if (query_line[0] == '>')
  {
    query.description = (char*) xmalloc(len);
    strcpy(query.description, query_line+1);
    query.dlen = len-1;
    query_line[0] = 0;
    fgets(query_line, LINE_MAX, query_fp);
  }
  else
  {
    query.description = (char*) xmalloc(1);
    query.description[0] = 0;
    query.dlen = 0;
  }

  int size = LINE_MAX;
  char * query_sequence = (char *) xmalloc(size);
  query_sequence[0] = 0;
  long query_length = 0;
  char * query_sequence_unmasked = NULL;
  if (mask) {
    query_sequence_unmasked = (char *) xmalloc(size);
    query_sequence_unmasked[0] = 0;
  }
 
  char * map;
  char m;

  if (symtype == 5)
    map = map_sound;
  else if ((symtype == 1) || (symtype == 3))
    map = map_ncbi_aa;
  else
    map = map_ncbi_nt16;

  while(query_line[0] && (query_line[0] != '>'))
  {
    char * p = query_line;
    while(int c = *p++)
    {
      if ((m = map[c]) >= 0)
      {
	if (query_length + 1 >= size)
	{
	  size += LINE_MAX;
	  query_sequence = (char*) xrealloc(query_sequence, size);
      if (mask)
        query_sequence_unmasked = (char*) xrealloc(query_sequence_unmasked, size);
	}
    if (mask) {
	  query_sequence_unmasked[query_length] = m;
	  if (mask && islower(c)) {
        if ((symtype == 1) || (symtype == 3))
          m = map['X'];
        else if (symtype != 5) // don't know what the wildcard character is for the sound alphabet (if existing at all)
          m = map['N'];
      }
    }
	query_sequence[query_length++] = m;
      }
    }
    query_line[0] = 0;
    fgets(query_line, LINE_MAX, query_fp);
  }
  query_sequence[query_length] = 0;
  if (mask)
    query_sequence_unmasked[query_length] = 0;
    
  if ((symtype == 0) || (symtype == 2) || (symtype == 4))
  {
    if (mask)
      query.nt[0].seq_unmasked = query_sequence_unmasked;
    query.nt[0].seq = query_sequence;
    query.nt[0].len = query_length;

    if (query.strands & 2)
    {
      //      printf("Reverse complement.\n");
      if (mask)
        query.nt[1].seq_unmasked = revcompl(query.nt[0].seq_unmasked, query.nt[0].len);
      query.nt[1].seq = revcompl(query.nt[0].seq, query.nt[0].len);
      query.nt[1].len = query.nt[0].len;
    }
    
    if ((symtype == 2) || (symtype == 4))
    {
      for(long s=0; s<2; s++)
      {
	if ((s+1) & query.strands)
	{
	  for(long f=0; f<3; f++)
	  {
	    translate(query.nt[0].seq, query.nt[0].len, s, f, 0,
		      & query.aa[3*s+f].seq, & query.aa[3*s+f].len);
        if (mask)
          translate(query.nt[0].seq_unmasked, query.nt[0].len, s, f, 0,
                & query.aa[3*s+f].seq_unmasked, & query.aa[3*s+f].len);
	  }
	}
      }
    }
  }
  else
  {
    if (mask) {
      query.aa[0].seq_unmasked = query_sequence_unmasked;
      Blast_ReadAaComposition(&query.composition_unmasked, BLASTAA_SIZE, (const Uint1*)query_sequence_unmasked, query_length);
    }
    query.aa[0].seq = query_sequence;
    query.aa[0].len = query_length;
    Blast_ReadAaComposition(&query.composition, BLASTAA_SIZE, (const Uint1*)query_sequence, query_length);
  }

  return 1;
}

char * revcompl(char * seq, long len)
{
  char * rc = (char *) xmalloc(len+1);
  for(long i=0; i<len; i++)
    rc[i] = ntcompl[(int)(seq[len-1-i])];
  rc[len] = 0;
  return rc;
}

void translate_createtable(long tableno, char * table)
{
  /* initialize translation table */

  for(long a=0; a<16; a++)
    for(long b=0; b<16; b++)
      for(long c=0; c<16; c++)
      {
	char aa = '-';
	for(long i=0; i<4; i++)
	  for(long j=0; j<4; j++)
	    for(long k=0; k<4; k++)
	    {
	      if ((a & (1<<i)) && (b & (1<<j)) && (c & (1<<k)))
	      {
		long codon = remap[i]*16 + remap[j]*4 + remap[k];
		char x = code[tableno-1][codon];
		if (aa == '-')
		{
		  aa = x;
		}
		else if (aa == x)
		{
		}
		else if ((aa == 'B') && ((x == 'D') || (x == 'N')))
		{
		}
		else if ((aa == 'D') && ((x == 'B') || (x == 'N')))
		{
		  aa = 'B';
		}
		else if ((aa == 'N') && ((x == 'B') || (x == 'D')))
		{
		  aa = 'B';
		}
		else if ((aa == 'Z') && ((x == 'Q') || (x == 'E')))
		{
		}
		else if ((aa == 'E') && ((x == 'Z') || (x == 'Q')))
		{
		  aa = 'Z';
		}
		else if ((aa == 'Q') && ((x == 'Z') || (x == 'E')))
		{
		  aa = 'Z';
		}
		else
		{
		  aa = 'X';
		}
	      }
	    }
	
	if (aa == '-')
	  aa = 'X';

	table[256*a+16*b+c] = map_ncbi_aa[(int)aa];
      }

#if 0
  /* dump it */
  
  printf("          -ACMGRSVTWYHKDBN\n");
  for(long x=0; x<16; x++)
    for(long y=0; y<16; y++)
    {
      printf("%2ld %2ld %c %c ", x, y, sym_ncbi_nt16[x], sym_ncbi_nt16[y]);
      for(long z=0; z<16; z++)
      {
	printf("%c", sym_ncbi_aa[table[256*x+16*y+z]]);
      }
      printf("\n");
    }
#endif
}

void translate_init(long qtableno, long dtableno)
{
  translate_createtable(qtableno, q_translate);
  translate_createtable(dtableno, d_translate);
}

void translate(char * dna, long dlen, 
	       long strand, long frame, long table,
	       char ** protp, long * plenp)
{
  //  printf("dlen=%ld, strand=%ld, frame=%ld\n", dlen, strand, frame);

  char * ttable;
  if (table == 0)
    ttable = q_translate;
  else
    ttable = d_translate;

  long pos, c;
  long ppos = 0;
  long plen = (dlen - frame) / 3;
  char * prot = (char*) xmalloc(1+plen);

  if (strand == 0)
  {
    pos = frame;
    while(ppos < plen)
    {
      c = dna[pos++];
      c <<= 4;
      c |= dna[pos++];
      c <<= 4;
      c |= dna[pos++];
      prot[ppos++] = ttable[c];
    }
  }
  else
  {
    pos = dlen - 1 - frame;
    while(ppos < plen)
    {
      c = ntcompl[(int)(dna[pos--])];
      c <<= 4;
      c |= ntcompl[(int)(dna[pos--])];
      c <<= 4;
      c |= ntcompl[(int)(dna[pos--])];
      prot[ppos++] = ttable[c];
    }
  }

  prot[ppos] = 0;
  *protp = prot;
  *plenp = plen;
}


void query_show()
{
  int linewidth = 60;
  for (unsigned i=0; i<strlen(query.description); i+=linewidth)
  {
    if (i==0)
      fprintf(out, "Query description: %-60.60s\n", query.description+i);
    else
      fprintf(out, "                   %-60.60s\n", query.description+i);
  }

#if 0
  long qlen;
  char * qseq;
  if ((symtype == 0) || (symtype == 2) || (symtype == 4))
  {
    qseq = query.nt[0].seq;
    qlen = query.nt[0].len;
  }
  else
  {
    qseq = query.aa[0].seq;
    qlen = query.aa[0].len;
  }

  for (int j=0; j<qlen; j+=linewidth)
  {
    if (j==0)
      fprintf(out, "Query sequence:    ");
    else
      fprintf(out, "                   ");

    for(int k=0; k<linewidth; k++)
    {
      if (j+k < qlen)
	putc(query.sym[qseq[j+k]], out);
      else
	break;
    }
    fprintf(out, "\n");
  }
#endif
}

